File: packages/server/src/services/protocols.service.ts
import { AuditInfo } from '../shared';
import { RawProtocol } from '../api/defillama.protocols.api.service';
import { baseSlug } from '../utils/slug.utils';

export type ProtocolAuditMap = Map<string, AuditInfo>;

export function buildProtocolAuditMap(protocols: RawProtocol[]): ProtocolAuditMap {
  const map: ProtocolAuditMap = new Map();

  for (const p of protocols) {
    if (!p.slug) continue;
    const info: AuditInfo = { audits: parseInt(p.audits ?? '0', 10) || 0, auditLinks: p.audit_links ?? [] };
    map.set(p.slug, info);

    const base = baseSlug(p.slug);
    if (base !== p.slug && !map.has(base)) {
      map.set(base, info);
    }
  }

  return map;
}

/** Return audit info for a given project slug, or null if not found. */
export function matchAuditInfo(projectSlug: string, protocolMap: ProtocolAuditMap): AuditInfo | null {
  if (protocolMap.has(projectSlug)) return protocolMap.get(projectSlug)!;

  const base = baseSlug(projectSlug);
  if (base !== projectSlug && protocolMap.has(base)) return protocolMap.get(base)!;

  return null;
}

File: src/Credfeto.Defi.Services/ProtocolsService.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using Credfeto.Defi.Data.Models.Models;
using Credfeto.Defi.Services.Utils;

namespace Credfeto.Defi.Services;

/// <summary>
///     Builds and queries a slug-keyed map of protocol audit information.
/// </summary>
public static class ProtocolsService
{
    /// <summary>
    ///     Builds a map from protocol slug to audit information.
    /// </summary>
    public static IReadOnlyDictionary<string, AuditInfo> BuildProtocolAuditMap(IReadOnlyList<RawProtocol> protocols)
    {
        Dictionary<string, AuditInfo> map = new(StringComparer.OrdinalIgnoreCase);

        foreach (RawProtocol p in protocols)
        {
            if (string.IsNullOrEmpty(p.Slug))
            {
                continue;
            }

            AuditInfo info = new()
            {
                Audits = int.TryParse(
                    s: p.Audits,
                    style: NumberStyles.Integer,
                    provider: CultureInfo.InvariantCulture,
                    result: out int count
                )
                    ? count
                    : 0,
                AuditLinks = p.AuditLinks ?? [],
            };

            _ = map.TryAdd(key: p.Slug, value: info);

            string baseSlug = SlugUtils.BaseSlug(p.Slug);

            if (!string.Equals(a: baseSlug, b: p.Slug, comparisonType: StringComparison.OrdinalIgnoreCase))
            {
                _ = map.TryAdd(key: baseSlug, value: info);
            }
        }

        return map;
    }

    /// <summary>
    ///     Returns audit info for a given project slug, or null if not found.
    /// </summary>
    public static AuditInfo? MatchAuditInfo(string projectSlug, IReadOnlyDictionary<string, AuditInfo> protocolMap)
    {
        if (protocolMap.TryGetValue(key: projectSlug, out AuditInfo? info))
        {
            return info;
        }

        string baseSlug = SlugUtils.BaseSlug(projectSlug);

        return
            !string.Equals(a: baseSlug, b: projectSlug, comparisonType: StringComparison.OrdinalIgnoreCase)
            && protocolMap.TryGetValue(key: baseSlug, out AuditInfo? baseInfo)
            ? baseInfo
            : null;
    }
}
